[CSSimplify] Detect and diagnose generic argument mismatches individually

xedin opened the pull request:
Generic arguments types are not always resolved enough to enable
aggregated mismatch fixes, which means that the solver should be
able to handle standalone generic argument matching constraints
and create a fix per mismatch location to coalesce them during
diagnostics.

Resolves: rdar://106054263

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ConstraintSystem] NFC: Pass down matcher flags to `repairFailures`

xedin committed:
[CSFix] Implement coalescing for generic argument mismatch fixes

xedin committed:
[CSSimplify] Detect and diagnose generic argument mismatches individually

Generic arguments types are not always resolved enough to enable
aggregated mismatch fixes, which means that the solver should be
able to handle standalone generic argument matching constraints
and create a fix per mismatch location to coalesce them during
diagnostics.

Resolves: rdar://106054263

xedin commented:
@swift-ci please test

LucianoPAlmeida commented on include/swift/Sema/ConstraintLocatorPathElts.def:
```diff
@@ -181,6 +181,10 @@ CUSTOM_LOCATOR_PATH_ELT(SynthesizedArgument)
 CUSTOM_LOCATOR_PATH_ELT(TupleType)
 
+/// A generic type, which provides context for subsequent generic
+/// argument pathc components.
```
Typo? `pathc` -> `path`

xedin replied:
Oops, thank you!

xedin replied:
✅

xedin committed:
[ConstraintSystem] Add a locator element to represent a generic type

This locator is going to be used as a "parent" element for
"generic argument" elements.

LucianoPAlmeida commented on lib/Sema/CSFix.cpp:
```diff
@@ -732,10 +732,34 @@ AllowFunctionTypeMismatch::create(ConstraintSystem &cs, Type lhs, Type rhs,
+
+  SmallVector<unsigned> mismatches(scratch.begin(), scratch.end());
+  // Make sure that notes about generic arguments are always produced in order.
+  llvm::array_pod_sort(mismatches.begin(), mismatches.end());
```
Not that it matters much in this case, but do we need to sort? `std::set` is a sorted container on `Compare` which defaults is `class Compare =std::less<Key>`, so `mismatches` would be constructed sorted already since is begin to end.

xedin replied:
That’s what I read as well but for some reason one of the tests produced notes out of order and I decided to just sort always to make sure there is no flakiness.

xedin replied:
Maybe there was something else going on while I was testing though, will give it a try without sort a couple of times to make sure…

LucianoPAlmeida replied:
> Maybe there was something else going on while I was testing though, will give it a try without sort a couple of times to make sure…

No worries, I was just curious =]
Even looked for a bit in the `SmallVector` [implementation](https://llvm.org/doxygen/SmallVector_8h_source.html#l01215) to see if the order may be somehow lost there but it seems like not.

xedin replied:
Let's see if we can reproduce this, I might have been using `SmallSet` when I ran the tests...

xedin replied:
Looks like it’s fine :)

xedin commented:
@swift-ci please test

hborla commented on test/Constraints/generics.swift:
```diff
@@ -1031,3 +1031,27 @@ func test_requirement_failures_in_ambiguous_context() {
+// rdar://106054263 - failed to produce a diagnostic upon generic argument mismatch
+func test_mismatches_with_dependent_member_generic_arguments() {
+  struct Binding<T, U> {}
+  // expected-note@-1 {{arguments to generic parameter 'T' ('Double?' and 'Data.SomeAssociated') are expected to be equal}}
```
Cool! Is it not possible to simplify the dependent member types in these cases?

xedin replied:
Diagnostic engine does (aka) but not very consistently…

hborla replied:
Yeah I've noticed the same thing. Well, we can look into that separately!

xedin replied:
Sounds good to me!

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@swift-ci please test Linux platform
